miden-lib/asm/sat/internal/prologue.masm

```diff
@@ -133,6 +134,91 @@ end
 # ACCOUNT DATA
 # =================================================================================================
 
+#! Ingests the account storage slot type data and stores it in memory at the appropriate memory
+#! addresses.
+#!
+#! This is achieved by reading the storage type commitment from storage slot 255 and then reading
+#! the storage type data from the advice provider.
+#!
+#! Stack: []
+#! Merkle Store: { ACCT_STORAGE[slot_255] -> TYPES_COM }
+#! Advice Map: { TYPES_COM: [slot_0_type, slot_1_type, ..., slot_255_type] }
+#! Output: []
+proc.ingest_acct_storage_types
+    # get the storage slot types commitment
+    exec.account::get_slot_types_commitment_storage_slot exec.account::get_item
+    # => [TYPES_COM]
+
+    # load the type data from the advice map onto the advice stack
+    # we prefix with [push.1 drop] due to https://github.com/0xPolygonMiden/miden-vm/issues/1122
+    push.1 drop adv.push_mapval
+    # => [TYPES_COM]
+
+    # get the storage slot types data pointer
+    exec.layout::get_acct_storage_slot_type_data_ptr
+    # => [slot_type_data_ptr, TYPES_COM]
+
+    # prepare the stack for reading storage slot types
+    padw padw padw
+    # => [ZERO, ZERO, ZERO, slot_type_data_ptr, TYPES_COM]
+
+    # read the storage slot types from the advice provider and store in memory
+    repeat.32
+        adv_pipe hperm
+        # => [PERM, PERM, PERM, slot_type_data_ptr', TYPES_COM]
+    end
+
+    # extract the digest and drop the memory pointer
+    dropw swapw dropw movup.4 drop
+    # => [DIGEST, TYPES_COM]
+
+    # assert the digest matches the commitment
+    assert_eqw
+    # => []
+end
+
+#! Validates that storage slot types are well formed.
+#!
+#! This is achieved by reading the storage slot types from memory and asserting that they are
+#! well formed. Currently we only support scalar slots with an entry of maximum arity of 256.
+#!
+#! Stack: []
+#! Output: []
+proc.validate_storage_slot_types
```

Comment:
The validation of storage slot types only needs to be performed once upon account creation. However, the ingestion of account storage types needs to be performed for every transaction. Having them decoupled means we do not perform the redundant type validation for every transaction. Validation of slot types costs ~ 255 * 41 = 10455 cycles to the transaction kernel. I think it would preferable to only do this upon account creation and not for every transaction. 

We could put the validation logic in the ingestion procedure behind an `if` block (conditional on whether the account is new) but I've tried to keep all logic associated with new account validation within the `validate_new_account` procedure for isolation and hygiene reasons.

What do you think?

Reply:
Ah yes - makes sense. Let's keep them separated as you currently have. (I left a separate comment with which I believe we can bring the cycle count down closer to ~6K).